### PR TITLE
Implement UART type erasure

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### CODE REVIEW

**Review by Dominaezzz — APPROVED** [review]

**Review by bjoernQ — APPROVED** [review]
LGTM - feel free to think about doc-hiding the Instance trait or merge as is

File: esp-hal/src/uart.rs [hunk 1]
```diff
@@ -1274,122 +1379,117 @@ where
...
+    fn uart_number(&self) -> usize;
+
+    /// Returns the interrupt handler for this UART instance.
+    fn async_handler(&self) -> InterruptHandler;
 
     /// Returns the interrupt associated with this UART instance.
-    fn interrupt() -> Interrupt;
+    fn interrupt(&self) -> Interrupt;
+
+    #[doc(hidden)]
```
[the remark this answers is not in the document]
Comment (bugadani) [reply esp-hal/src/uart.rs:1454]: In the long run, I want to split out the RegisterBlock/signal part of the trait, and make it public. The rest (i.e. the actual driver implementations) will I think remain private. I'm hiding these things mostly as an "under construction" sign.

**Comment by bugadani** [comment]
Are we fine with me disabling LTO in the test suite?

**Comment by bjoernQ** [comment]
> Are we fine with me disabling LTO in the test suite?

What would be the alternative?

**Pull request merged**